### PR TITLE
Implement `axis.layer` with 'above traces' and 'below traces' values

### PR DESCRIPTION
**Pull request by etpinard**
to be merged in https://github.com/plotly/plotly.js/pull/1861 with (hopefully) all the requirements for https://github.com/plotly/plotly.js/issues/1385

This PR adds an axis attribute `layer` with two possible values: `'above traces'` (the default and current behavior) and `'below traces`' which allows together with `cliponaxis: false` to draw markers and text nodes above axis lines and tick labels. See `cliponaxis_false` mock in e44aa4d for an example.

We chose to make `layer` have a _hard_ `'above traces'` default value corresponding to the behavior on `master`. This might be revisited in [`v2`](https://github.com/plotly/plotly.js/issues/420).

This PR also fixes a bug in the `cliponaxis: false` implementation in 7b62b73 and lints a few things.

-----

In a private convo with @alexcjohnson , we opted to **not** include special auto-range logic for `cliponaxis: false` traces. This issue will be dealt with by having a few more axis options in a later PR. I'll open up a ticket about this tomorrow. **DONE** in https://github.com/plotly/plotly.js/issues/1876

Unless someone :no_entry: me, <strike>I'll won't add a workaround patch for the cartesian line clip path problems on pan (see https://github.com/plotly/plotly.js/pull/1861#issuecomment-313755166) nor will I fix the IE / Edge issues from https://github.com/plotly/plotly.js/pull/1861#issuecomment-313755166. In trying to fix these, I broke some animation tests; I think it would be wise to wait until Ricky is back to working full-time before making another attempt. I'll document this things in new separate issues tomorrow.</strike> **FIXED** in e234827

### CODE REVIEW

**Comment by alexcjohnson** [comment]
> I won't add a workaround patch for the cartesian line clip path problems on pan (see #1861 (comment)) nor will I fix the IE / Edge issues from #1861 (comment).

That's fine, I'll make an issue for the IE/Edge one since it's not new. For the chrome issue since it IS new, can you investigate a little (perhaps reduce it to a static SVG bug in codepen?) and then either make an internal issue or report it to chromium?

File: test/jasmine/tests/cartesian_test.js [hunk 1]
```diff
@@ -403,4 +403,79 @@ describe('subplot creation / deletion:', function() {
...
+            _assert(
+                [true, 10],
+                [true, 10],
+                [false, 0],
+                [false, 0]
+            );
+        })
+        .catch(failTest)
+        .then(done);
+    });
```
Comment (alexcjohnson) [inline test/jasmine/tests/cartesian_test.js]: Beautiful. This test is exactly the question I was about to ask when looking at the code.

File: src/plots/cartesian/layout_attributes.js [hunk 2]
```diff
@@ -573,6 +573,18 @@ module.exports = {
...
+    layer: {
+        valType: 'enumerated',
+        values: ['above traces', 'below traces'],
+        dflt: 'above traces',
+        role: 'info',
+        description: [
+            'Sets the layer on which this axis is displayed.',
+            'If *above traces*, this axis is displayed above all the subplot\'s traces',
+            'If *below traces*, this axis is displayed below all the subplot\'s traces,',
+            'but above the grid lines.'
```
Comment (alexcjohnson) [inline src/plots/cartesian/layout_attributes.js]: Per https://github.com/plotly/plotly.js/pull/1861#issuecomment-314523923 - can you mention `scatter.cliponaxis` here (and vice versa)?

Comment (etpinard) [inline src/plots/cartesian/layout_attributes.js]: Good call :books:

File: tasks/util/strict_d3.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+var pathToStrictD3Module = path.join(
+    constants.pathToImageTest,
+    'strict-d3.js'
+);
+
+/**
+ * Transform `require('d3')` expressions to `require(/path/to/strict-d3.js)`
+ */
+
+module.exports = transformTools.makeRequireTransform('requireTransform',
```
Comment (etpinard) [inline tasks/util/strict_d3.js]: @alexcjohnson I hope you don't mind this commit too much. This made incorporating `strict-d3.js` in our jasmine tests _a lot_ easier - which is very useful e.g. to :lock: down https://github.com/plotly/plotly.js/pull/1871/commits/e234827dac252743bb68329487cabaee80dab953

Companion `plotly-mock-viewer` PR: https://github.com/rreusser/plotly-mock-viewer/pull/15

File: test/image/strict-d3.js [hunk 4]
```diff
@@ -25,6 +41,14 @@ selProto.style = function() {
     return originalSelStyle.apply(sel, arguments);
 };
 
+function checkAttrVal(sel, key) {
+    // setting the transform attribute on a <clipPath> does not
+    // work in Chrome, IE and Edge
+    if(sel.node().nodeName === 'clipPath' && key === 'transform') {
```
Comment (etpinard) [inline test/image/strict-d3.js]: No need for additional jasmine tests cases, this here made many animation and interaction test fail.

File: src/plots/cartesian/dragbox.js [hunk 5]
```diff
@@ -743,7 +743,7 @@ module.exports = function dragBox(gd, plotinfo, x, y, w, h, ns, ew) {
             var plotDx = xa2._offset - clipDx / xScaleFactor2,
                 plotDy = ya2._offset - clipDy / yScaleFactor2;
 
-            fullLayout._defs.selectAll('#' + subplot.clipId)
+            fullLayout._defs.select('#' + subplot.clipId + '> rect')
```
Comment (etpinard) [inline src/plots/cartesian/dragbox.js]: See https://codepen.io/etpinard/pen/eRbxrp?editors=1010 for more info on this topic.

File: tasks/util/strict_d3.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    { evaluateArguments: true, jsFilesOnly: true },
+    function(args, opts, cb) {
+        var pathIn = args[0];
+        var pathOut;
+
+        if(pathIn === 'd3' && opts.file !== pathToStrictD3Module) {
```
Comment (alexcjohnson) [inline tasks/util/strict_d3.js]: I gather this is so that strict-d3 itself can still get the original d3?

File: tasks/util/strict_d3.js (hunk 3 above)
Comment (alexcjohnson) [inline tasks/util/strict_d3.js]: yeah, looks great!

File: tasks/util/strict_d3.js (hunk 6 above)
Comment (etpinard) [inline tasks/util/strict_d3.js]: Bingo.

File: src/plots/cartesian/layout_attributes.js (hunk 2 above)
Comment (etpinard) [inline src/plots/cartesian/layout_attributes.js]: done in 6f494c692f70c8d2714d1c4457b9dbbd5da2eb90

**Comment by alexcjohnson** [comment]
💃

**Pull request merged**